etl/churn_to_csv.kp/knowledge.md

```diff
@@ -0,0 +1,143 @@
+
+    # upload files to s3
+    try: 
+        for config in upload_config:
+            print("{}: Uploading to {} at s3://{}/{}/{}".format(
+                    datetime.utcnow(), config["name"], config["bucket"], 
+                    config["prefix"], filename))
+
+            s3_path = "{}/{}".format(config["prefix"], filename)
+            transfer.upload_file(filename, config["bucket"], s3_path,
+                                 extra_args={'ACL': 'bucket-owner-full-control'})
+    except botocore.exceptions.ClientError as e:
+        print("File for {} already exists, skipping upload: {}".format(filename, e))
+
+
+def marginalize_dataframe(df):
+    """ Reduce the granularity of the dataset to the original set of attributes.
+    The original set of attributes can be found on commit 2de3ef1 of mozilla-reports. """
+    
+    attributes = ['channel', 'geo', 'is_funnelcake',
+                  'acquisition_period', 'start_version', 'sync_usage',
+                  'current_version', 'current_week', 'is_active']
+    aggregates = ['n_profiles', 'usage_hours', 'sum_squared_usage_hours']
+    
+    return df.groupby(attributes).agg(*[F.sum(x).alias(x) for x in aggregates])
+
+
+def convert_week(config, week_start=None):
+    """ Convert a given retention period from parquet to csv. """
+    df = spark.read.parquet(config["source"])
+    
+    # find the latest start date based on the dataset if not provided
+    if not week_start:
+        start_dates = df.select("week_start").distinct().collect()
+        week_start = sorted(start_dates)[-1].week_start
+    
+    # find the week end for the filename
+    dateformat = "%Y%m%d"
+    week_end = (
+        datetime.strptime(week_start, dateformat) + timedelta(6)
+    ).strftime(dateformat)
+    
+    print("Running for the week of {} to {}".format(week_start, week_end))
+    
+    # find the target subset of data
+    df = df.where(df.week_start == week_start)
+    
+    # marginalize the dataframe to the original attributes and upload to s3
+    upload_df = marginalize_dataframe(df)
+    filename = "churn-{}-{}.by_activity.csv.gz".format(week_start, week_end)
+    collect_and_upload_csv(upload_df, filename, config["uploads"])
+
+```
+
+```python
+def assert_valid_config(config):
+    """ Assert that the configuration looks correct. """
+    assert set(["source", "uploads"]).issubset(config.keys())
+    for entry in config["uploads"]:
+        assert set(["name", "bucket", "prefix"]).issubset(entry.keys())
+```
+
+```python
+config = {
+    "source": "s3://telemetry-parquet/churn/v2",
+    "uploads": [
+        {
+            "name":   "Pipeline-Analysis",
+            "bucket": "net-mozaws-prod-us-west-2-pipeline-analysis",
+            "prefix": "mreid/churn"
+        },
+        {
+            "name":   "Dashboard",
+            "bucket": "net-mozaws-prod-metrics-data",
+            "prefix": "telemetry-churn"
+        }
+    ]
+}
+
+if False:
```

Comment:
Are you planning to leave this part in the notebook? If so, please leave a comment here stating that's just for debugging purposes and it's disabled.